avalanche/benchmarks/utils/data.py

```diff
@@ -0,0 +1,714 @@
+    applied to training patterns. Those transformations usually involve some
+    kind of data augmentation. The second one is 'eval', that will contain
+    transformations applied to test patterns. Having both groups can be
+    useful when, for instance, in need to test on the training data (as this
+    process usually involves removing data augmentation operations). Switching
+    between transformations can be easily achieved by using the
+    :func:`train` and :func:`eval` methods.
+
+    Moreover, arbitrary transformation groups can be added and used. For more
+    info see the constructor and the :func:`with_transforms` method.
+    AvalancheDataset can switch between the 'train' and
+    'eval' groups by calling the ``train()`` and ``eval()`` methods. When
+    using custom groups one can use the ``with_transforms(group_name)``
+    method instead.
+
+    This dataset will try to inherit the task labels from the input
+    dataset. If none are available and none are given via the `task_labels`
+    parameter, each pattern will be assigned a default task label "0".
+    See the constructor for more details.
+    """
+
+    def __init__(
+        self,
+        dataset: IDataset,
+        *,
+        data_attributes: List[DataAttribute] = None,
+        transform_groups: TransformGroups = None,
+        frozen_transform_groups: TransformGroups = None,
+        collate_fn: Callable[[List], Any] = None
+    ):
+        """Creates a ``AvalancheDataset`` instance.
+
+        :param dataset: Original dataset. Beware that
+            AvalancheDataset will not overwrite transformations already
+            applied by this dataset.
+        :param transform_groups: Avalanche transform groups.
+        """
+        # NOTES on implementation:
+        # - raw datasets operations are implemented by _FlatData
+        # - data attributes are implemented by DataAttribute
+        # - transformations are implemented by TransformGroups
+        # AvalancheDataset just takes care to manage all of these attributes
+        # together and decides how the information propagates through
+        # operations (e.g. how to pass attributes after concat/subset
+        # operations).
+
+        # original dataset. Don't use this attribute directly because some subclasses
+        # (e.g. AvalancheConcatDataset) don't have it.
+        # This attribute is a list to make it compatible with AvalancheConcatDataset
+        self._dataset = dataset
+        self._data_attributes = {}
+
+        ####################################
+        # Init data attributes
+        ####################################
+        if isinstance(dataset, AvalancheDataset):
+            # inherit data attributes from original dataset
+            self._data_attributes = {**dataset._data_attributes}
+
+        if data_attributes is not None:
+            da_dict = {da.name: da for da in data_attributes}
+            self._data_attributes.update(da_dict)
```

Comment:
I think that a way to add dataset fields as data attributes by specifying their name (using the attribute with the same name of the dataset object) may be useful. Is it doable?

Reply:
Datasets are designed to be immutable objects. Changing attributes should be done by wrapping existing datasets e.g. `new_data=AvalancheDataset(old_data, data_attributes=[...])`. If we allow mutability indexing breaks.

Reply:
Sorry, I put this in the wrong way. What I was trying to say is that, when wrapping a non-Avalanche dataset that contains a useful field, that field could be specified like this:

```
plain_dataset = ...
avl_dataset = AvalancheDataset(plain_dataset, data_attributes=['some_field', ...])
```

instead of:
```
plain_dataset = ...
attributes = [DataAttribute(plain_dataset.some_field, name='some_field'), ...]
avl_dataset = AvalancheDataset(plain_dataset, data_attributes=attributes)
```

just to improve usability a little bit.